### ISSUE TITLE
Dynamic Rendering Feature (performance improvement)

### ISSUE THREAD

**Issue by egmracer01** [ianstormtaylor/slate#790]
#### Do you want to request a *feature* or report a *bug*?
Feature. 

The proposed idea is to dynamically render content within slate. This means that slate would only render visible blocks and not render blocks hidden within the y-overflow. The benefits that this would provide for performance are huge. By far the slowest part of the initial render is mounting all of the components (if I'm reading the timeline correctly). You could effectively bound the number of dom nodes rendered at any given time and perhaps find other optimizations as a result.

<img width="1318" alt="screen shot 2017-05-04 at 11 06 20 pm" src="https://cloud.githubusercontent.com/assets/1102811/25735189/61d07b60-311e-11e7-9e6c-e73c4e03557c.png">

[slate-large.zip](https://github.com/ianstormtaylor/slate/files/978215/slate-large.zip)

#### What's the current behavior?
All of the components are rendered.

#### What's the desired behavior?
Only visible (with a given padding of elements to support smooth scrolling

The Ace Editor: https://github.com/ajaxorg/ace
Does this by rendering two divs: a full height "scroller" div and then a "content" div that dynamic adjusts based on scrolls while updating it's content (removing hidden dom nodes and adding newly shown dom nodes). 

Here's how the position of the content div updates: https://github.com/ajaxorg/ace/blob/master/lib/ace/virtual_renderer.js#L838
Here's how it calculates the properties of the content div: https://github.com/ajaxorg/ace/blob/master/lib/ace/virtual_renderer.js#L1025

Here's an example of dynamic rendering in react: https://github.com/clauderic/react-tiny-virtual-list

**Comment by optimistiks** [comment]
Interesting. I wonder how it's possible to preserve capability of selecting everything with Ctrl+A with that feature. It seems that in Ace editor all features are implemented from scratch (caret, selection, etc), and Slate is using contenteditable, so it may be more difficult to implement such feature in Slate.

**Comment by egmracer01** [comment]
You could preserve Ctrl+A pretty easily by handling that case specifically. You could have a virtual selection and then a rendered selection. I think the larger question is to what degree would you have to implement features and handle edge cases.

This would seem to be a significant change but I think you would be able to handle most of the caret & selection with offsets.

**Comment by ianstormtaylor** [comment]
Hey @egmracer01 I think this sounds very interesting, going to leave it open for people to discuss how it might be done. I haven't done any virtualization with React before, so I have nothing to contribute right now.

I'd be curious to know which pieces of the rendering could be avoided with virtualization. i think some of that logic might not be avoided, and there might be lowering-hanging fruit elsewhere?

**Comment by egmracer01** [comment]
one piece of lower hanging fruit might be related to: https://medium.com/missive-app/45-faster-react-functional-components-now-3509a668e69f . To summarize the problem, pure rendering components are simply wrappers around normal components and follow the same code paths. By calling the function `Component(props)` instead of `<Component {...props} />` you can get a big speed improvement without consequences. This appears to be something that the react team is looking into already.

From my understanding, nodes with custom types are pure: for example: https://github.com/ianstormtaylor/slate/blob/master/examples/rich-text/index.js#L20 . 

This means that when we render these components: https://github.com/ianstormtaylor/slate/blob/17ea3ed349963e91454b3d788a5242036565ab55/src/components/node.js#L272 assuming `Component` is pure, we actually could speed this up by changing it to something like: 
```
Component({
        attributes,
        key: node.key,
        editor,
        parent,
        node,
        readOnly,
        state,
        children
})
```

**Comment by ms88privat** [comment]
I would wait for React 0.16.0 and then have a new evaluation on it.

**Comment by thesunny** [comment]
I just wanted to chime in here with a few points on what I believe is and isn't possible and a really good low hanging fruit IMO.

I don't think it's possible to arbitrarily render any part of a Slate Document at a given scroll position because it is difficult, possibly impossible, to tell how much space a Slate block will take until after it is rendered. Ace editor is an editor for monospace text so it is easy to predict how much space text will take before we render it.

That said, what could work and could improve performance, is to render the visible portion of the page first before giving control back to the user. The rest of the content could fill in afterwards.

I think this might be a good low hanging fruit. Just render up to the visible portion, then use setTimeout to render a few blocks at a time for the rest without blocking the UI.

This would give the appearance of being usable quickly on very long documents. The only thing you'd see is the scrollbar getting longer.

**Comment by ianstormtaylor** [comment]
@thesunny good points!

For anyone needing this, I would highly recommend looking into other performance improvements in the general rendering case first, because they are probably much lower-hanging fruit.

**Comment by linonetwo** [comment]
https://github.com/bvaughn/react-window Might be a solution to this, but it requires you to hand over the rendering to it:

```jsx
/// fake example
import { VariableSizeList as List } from 'react-window';

function AFastEditor() {
  const getItemSize = index => editor.blocks[index].getHeight();

  // style is used to absolutely locate elements in the long parent
  const getBlock = ({ index, style }) => React.cloneElement(editor.blocks[index], { style });

  return (
    <List
      height={150}
      itemCount={1000}
      itemSize={getItemSize}
      width={300}
    >
      {getBlock}
    </List>
  );
}
```

**Comment by linonetwo** [comment]
I tried to use `react-window` today, I hope the `children` I get from `renderEditor` is an array of rows to render, instead of Editor.

Or hope there can be a hook `renderContent`.

```jsx
import React from 'react';
import { Editor } from 'slate-react';
import { Value } from 'slate';
import Plain from 'slate-plain-serializer';
import { FixedSizeList } from 'react-window';
import AutoSIzer from 'react-virtualized-auto-sizer';

const windowlongTextPlugin = {
  renderEditor: (props, editor, next) => {
    const children = next();
    const RenderRow = ({ index, style }) => <div style={style}>{children[index]}</div>;
    console.log(children); // I hope children here is an array of rows to render, instead of Editor
    
    return (
      <FixedSizeList height={600} itemCount={children.length} itemSize={35} width={500}>
        {RenderRow}
      </FixedSizeList>
    );
  },
};

export default class HugeTextViewer extends React.Component<{ text?: string }> {
  public render() {
    return (
      <Editor
        defaultValue={Plain.deserialize('')}
        value={Plain.deserialize(this.props.text || '')}
        plugins={[windowlongTextPlugin]}
      />
    );
  }
}
```

**Comment by linonetwo** [comment]
Rendering is very fast, almost instantly, and `deserializeHugeText` is also fast. But before rendering, loading it to slate is slow.

<img width="1351" alt="屏幕快照 2019-06-26 下午2 03 15" src="https://user-images.githubusercontent.com/3746270/60154884-4a52f380-981b-11e9-9e0a-5621cb34366a.png">

Slate spends lots of time doing `normalize -> withoutNormalize -> setLength`:

<img width="833" alt="屏幕快照 2019-06-26 下午2 05 46" src="https://user-images.githubusercontent.com/3746270/60155002-94d47000-981b-11e9-92fa-09f0f16dea98.png">

**Comment by linonetwo** [comment]
![屏幕快照 2019-07-22 下午3 43 09](https://user-images.githubusercontent.com/3746270/61614940-76855700-ac97-11e9-8b82-2e2fe5c43fee.png)

`transformed.toArray()` takes a long time.

**Comment by pgsill** [comment]
@mmdonaldson I would absolutely love to see this. If the current Slate behaviors could be kept with deferred rendering it'd be fantastic.

**Comment by daldoy** [comment]
Is anyone working on this?

**Comment by shubham43MP** [comment]
Any Updates to efficiently render Long documents with slate editor here? I think this is important from the UX perspective as the editing experience might be compromised.

**Comment by AliMamed** [comment]
Hi all. I was experimenting with this at the `renderElement` prop of `Editable` in the following way: 
- use IntersectionObserver to look if element which is direct child of editor is close enough to the viewport 
- if not render a placeholder and don't render any children of this node
Not rendering of children breaks some workflows in `slate-react`. (Looks like it looks for those nodes in weak-maps and crashes when can't find them in DOM)

What is the community best practice on this? Is this feature still under consideration, or maybe I am missing some point?

**Comment by dylans** [comment]
> Hi all. I was experimenting with this at the `renderElement` prop of `Editable` in the following way:
> 
>     * use IntersectionObserver to look if element which is direct child of editor is close enough to the viewport
> 
>     * if not render a placeholder and don't render any children of this node
>       Not rendering of children breaks some workflows in `slate-react`. (Looks like it looks for those nodes in weak-maps and crashes when can't find them in DOM)
> 
> 
> What is the community best practice on this? Is this feature still under consideration, or maybe I am missing some point?

Still under consideration, but I don't think we've had a recent PR that would make this work. Would be happy to review or discuss further as it would be a nice improvement.

**Comment by andyjakubowski** [comment]
> I have a proof of concept. Is there still interest in this?

Yes, definitely. Optimizing performance for long Slate documents has been an ongoing challenge in a project I’m involved in. Would you mind sharing your approach @pau-not-paul? Your PoC looks great.
